Blocks: refactor deletion warnings dialog

glendaviesnz opened the pull request:
## What?
Refactors the block deletion warnings

## Why?
#58796 Added some code to show a warning if a user tries to delete a block that has pattern overrides, but could do with some code improvements.

This follow up makes the block removal warnings more flexible to cover the use case for patterns

## How?

- Some of the code (the block removal rules) is moved into the `wordpress/editor` package along with the copy, so that it works across the post and site editor automatically.
- The rules can specify a `postType` that the associated message should appear for. This is mostly use to ensure the message for pattern overrides only shows when editing a pattern (see [Patterns: Avoid showing block removal warning when deleting a pattern instance that has overrides](https://github.com/WordPress/gutenberg/pull/59044#top))
- Rather than hard coding a message, the rules now have a callback that receives the list of blocks being removed and returns a message to show. If no message is returned by any of the rules, no removal warning is displayed.
- I've made some small changes to the copy, but largely it's as it is in `trunk`.

## Testing Instructions

- Try deleting a post content block and query block in both the post and site editor and make sure a confirmation warning shows
- Add a synced pattern block and set overrides on one of the child blocks. Try deleting the child block and make sure the deletion warning message shows.

## Screenshots or screencast <!-- if applicable -->

![Screenshot 2024-02-21 at 4 16 53 pm](https://github.com/WordPress/gutenberg/assets/677833/1d91f231-20a5-4c72-983a-e37553b28ac6)

glendaviesnz committed:
Refactor block removal warnings to use a callback, and move them to the editor package

move the block specific message wording back into the editor provider so the block editor is not polluted with  `template` terminology, and move the pattern specific check into a callback which is passed in via the rules object

Move removal warnings setup into own component

Fix e2e

Another e2e fix

Remove the intro text

remove select from callback

Simplify the API so standard for specific blocks and more generic checks like pattern overrides

glendaviesnz commented:
@annezazu , @andrewserong [this PR](https://github.com/WordPress/gutenberg/pull/58138) introduced `template` specific language into the block editor component, which is not supposed to know about WP specific things like templates, so this PR pulls the block type specific wording out into the editor package. You may want to check the wording of the generic message and the block specific wording.

andrewserong commented on packages/block-editor/src/components/block-removal-warning-modal/index.js:
```diff
@@ -65,7 +52,25 @@ export function BlockRemovalWarningModal( { rules } ) {
 		>
-			<p>{ message }</p>
+			<div>
+				{ _n(
+					'Post or page content may not be displayed correctly if you delete this block:',
+					'Post or page content may not be displayed correctly if you delete these blocks:',
+					blockNamesForPrompt.length
+				) }
```
Since this PR is aiming to remove the dependency on language that the block editor package shouldn't know about, would it be worth making this a label param and have the consumer pass it in, rather than baking it into the component? Just wondering if `template` is too specific for the block editor package, then is post or page content similarly too WP specific? (I don't have a strong opinion here, the thought just occurred to me while reading through the change 🙂)

glendaviesnz replied:
see [this comment](https://github.com/WordPress/gutenberg/pull/58952#issuecomment-1940340129)

andrewserong reviewed (commented):
Thanks for the follow-up @glendaviesnz, good thinking seeing if we can remove the template-specific language from the block-editor package 👍

I notice this changes the wording a fair bit, so it'd be good to get some feedback on it from @annezazu or designers as I know they've done a few passes on tweaking the phrasing. To my eyes, it looks possibly a little long now:

### Before

![image](https://github.com/WordPress/gutenberg/assets/14988353/be644149-8dc1-43d2-9e83-6735d6fb690c)

### After

![image](https://github.com/WordPress/gutenberg/assets/14988353/30bf25d8-cefc-4118-813b-5f8c593f3fc3)

---

Also, just left a comment about the descriptive text added before the list of prompts. I was wondering if that could either be a param so that the consumer could pass it in, or potentially if we even need it? I.e. would the block-level warnings be enough on their own? 🤔

glendaviesnz commented:
> To my eyes, it looks possibly a little long now:

@andrewserong looking at it again, the intro text is actually a bit superfluous, so have removed it altogether this is how it looks now:

<img width="559" alt="Screenshot 2024-02-13 at 3 51 59 PM" src="https://github.com/WordPress/gutenberg/assets/3629020/bd6053eb-32c3-4f5b-b874-366b00cd8308">

<img width="552" alt="Screenshot 2024-02-13 at 3 52 15 PM" src="https://github.com/WordPress/gutenberg/assets/3629020/9d60e3ba-3dbd-43df-8bc3-c9173f25062f">

andrewserong commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,52 @@
+		message: __(
+			'Deleting a block with pattern instance overrides can break other blocks on your site that have content linked to it.'
+		),
+		callBack: ( select, clientId, blockNamesForPrompt ) => {
```
Passing the `select` across package boundaries in a `callback` feels a bit odd to me as an API. Is there a way we can avoid introducing the callback? I.e. would it be okay for the pattern overrides logic here to live in the component instead 🤔

glendaviesnz replied:
We were doing this in a block editor selector but it was suggested that this would be better [abstracted as callback in the rules](https://github.com/WordPress/gutenberg/pull/58796#discussion_r1483894219), to avoid complicating the block editor code with [block specific checks/filters](https://github.com/WordPress/gutenberg/pull/58796/files#r1486435447).

My reading of it wouldn't be that we are passing the `select` between package boundaries, it is just a callback that is passed into the block-editor package and the method is called within the `block-editor` package, so the `select` is not available for use in any way within the `editor` package. But happy to hear other's thoughts on this.

andrewserong replied:
Thanks Glen, I very well could have been overthinking this part! The main thought I had while looking at is is that I find callback signatures easier to comprehend when the params are fairly simple in nature (like with `clientId`, and `blockNamesForPrompt`), whereas `select` is a more powerful thing to be passed through since it's got all sorts of methods attached. So while the callback is executed in the `block-editor` package, the code in this bit of the `editor` package is doing stuff with a `select` where it isn't immediately obvious where the `select` came from.

It's definitely not a strongly held opinion, though, so happy to go with this if it's what folks prefer!

glendaviesnz replied:
we have time to think about this, so will take another look at how we might be able to refactor it without the `select`  - thanks for the feedback.

glendaviesnz replied:
@andrewserong I think we can simplify it down to [just passing in the block attributes](https://github.com/WordPress/gutenberg/pull/58952/commits/d0f24d2d543085e255e71df952676795a35a1ccf). 

We will just need to have a think about whether a third party might need something other than `blockAttributes` and `blockNamesForPrompt` to work out if they need to add a warning.

andrewserong replied:
Nice one. `blockAttributes` and `blockNamesForPrompt` sound like good params to start, to me!

andrewserong commented:
That's looking much better to me @glendaviesnz, let's see what the others think in terms of the change in wording.

The only issue from my perspective after re-looking at this is the passing around of the `select` across package boundaries, which felt a bit odd to me, as it mighn't make for a reliable API signature. I know it's a private component, but I was wondering if we can keep some of that logic contained still? I think ideally the `select` call would be happening on just one side of the boundary between packages, rather than being passed between 🤔

talldan commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,52 @@
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
+const blockRemovalRules = {
+	'core/query': {
+		message: __(
+			'The Query Loop block displays a list of posts or pages, so removing it will prevent that content from displaying.'
+		),
+	},
+	'core/post-content': {
+		message: __(
+			'Removing the Post Content block will stop your post or page content from displaying on this template.'
+		),
+	},
+	'core/post-template': {
+		message: __(
+			'The Post Template block displays each post or page in a Query Loop, so removing it will stop post content displaying in your query loop.'
+		),
+	},
```
A good reason to introduce a callback is to have some consistency in how the system works, but here there's still some blocks working one way (no callback) and then pattern overrides working another way (a callback).

Another problem with how it works is that some object keys are meaningful and represent one thing—`core/query`, `core/post-content` etc. are block names and are used for matching, while `core/pattern-overrides` is a block binding source and the key doesn't do anything.

As its all private code there's a chance to be a bit bolder, take a step back and consider how it all might work ideally to meet all the requirements.

glendaviesnz replied:
I was working from the point of view that it is reasonably standard for an API to have optional parameters and:
```js
const blockName = select.getBlockName( clientId );
if ( rules[ blockName ] ) {
	blockNamesForPrompt.add( blockName );
}
```
is already very generic so no point in passing it with every rule, so the `callBack` can just be optional for rules that require something less generic.

I see your point with the property naming though. 

There is no panic to get this change in, so we might as well take some time to think about the best approach.

glendaviesnz replied:
one option would be to separate block specific and generic rules and handle each group separately in the selector, eg.

```js
const blockSpecific = {
	'core/post-content': {
		message: __(
			'Removing the Post Content block will stop your post or page content from displaying on this template.'
		),
	},
};
const generic = {
	'core/pattern-overrides': {
		message: __(
			'Deleting a block with pattern instance overrides can break other blocks on your site that have content linked to it.'
		),
		callBack: ( select, clientId, blockNamesForPrompt ) => {
			const blockAttributes = select.getBlockAttributes( clientId );
			if (
				blockAttributes?.metadata?.bindings &&
				JSON.stringify( blockAttributes.metadata.bindings ).includes(
					'core/pattern-overrides'
				)
			) {
				blockNamesForPrompt.add( 'core/pattern-overrides' );
			}
		},
	},
};

const blockRemovalRules = { blockSpecific, generic };
```
with only the generic ones needing to provide a callback

glendaviesnz replied:
will have another think about various approaches/naming, etc. tomorrow.

glendaviesnz commented:
Have moved this back to draft as we need to take some time to think about the best API for the rules set, etc.

youknowriad commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,51 @@
+		message: __(
+			'The Query Loop block displays a list of posts or pages, so removing it will prevent that content from displaying.'
+		),
+	},
```
Are we showing this message for the removal of all query blocks?

glendaviesnz replied:
In trunk it is only shown on every deletion in the site editor for query and post content - which makes sense in the context of a template, but if we are adding these rules via the `editor` package I think we need to somehow limit their application to posts/pages - it doesn't seem like a warning is needed in the context of post and pages for these blocks and in fact would be confusing/annoying?

glendaviesnz replied:
have added a `postTypes` option to the rules to allow setting which post types to show the rules for. Current post type is passed in via the editor provider so the block editor does not need to be know anything about the different post types.

youknowriad commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,51 @@
+		message: __(
+			'Removing the Post Content block will stop your post or page content from displaying on this template.'
+		),
+	},
```
Are we showing this message for the removal of all post content blocks?

glendaviesnz replied:
see reply above.

youknowriad commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -131,27 +131,21 @@ export const privateRemoveBlocks =
-						blockNamesForPrompt.add( blockName );
-						messageType = 'patternOverrides';
+				Object.values( rules ).forEach( ( rule ) => {
+					if ( rule.callBack ) {
```
I believe in most of our APIs we used `callback` instead of `callBack`

glendaviesnz replied:
true, thanks, will fix that

youknowriad commented:
Thanks for the follow-up here :)

glendaviesnz commented:
we will need to account for [this fix](https://github.com/WordPress/gutenberg/pull/59044) in this refactor.

glendaviesnz commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,80 @@
+
+const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
```
@talldan I have revisited the rules to try and make them more generic, and also to add in the ability to limit the messages by post type. 

So now every rule needs:

key - this is the property name and is used to map to the messages in the confirmation prompt
    message - message to display
    postTypes - array of post types on which to display the message
    callback - method to check if warning should show - callback gets `blockName`, `ruleKeysForPrompt` and `blockAttributes` passed in

The advantage of just havng a callback for each is devs are then free to calculate this however they want either for specific blocks or more generic like the patterns.

This is just a first stab at this bigger refactor of this API so happy to take thoughts on alternatives.

glendaviesnz commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,80 @@
+const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
+const blockRemovalRules = {
+	'core-query': {
+		message: __(
+			'The Query Loop block displays a list of posts or pages, so removing it will prevent that content from displaying.'
+		),
+		postTypes: [ 'wp_template' ],
+		callback: ( blockName, ruleKeysForPrompt ) => {
+			if ( blockName === 'core/query' ) {
+				ruleKeysForPrompt.add( 'core-query' );
+			}
+		},
+	},
```
@talldan I have revisited the rules to try and make them more generic, and also to add in the ability to limit the messages by post type. 

So now every rule needs:

key - this is the property name and is used to map to the messages in the confirmation prompt
    message - message to display
    postTypes - array of post types on which to display the message
    callback - method to check if warning should show - callback gets `blockName`, `ruleKeysForPrompt` and `blockAttributes` passed in

Doing it this way makes for a bit of repetition for simple block checks, but the advantage of just havng a callback for each is devs are then free to calculate this however they want either for specific blocks or more generic things like the pattern overrides.

This is just a first stab at this bigger refactor of this API so happy to take thoughts on alternatives.

glendaviesnz replied:
we can probably make the individual block callback more generic and share it between the rules somehow if we decide to go  this way.

youknowriad commented on packages/block-editor/src/components/block-removal-warning-modal/index.js:
```diff
@@ -17,10 +17,10 @@ import { store as blockEditorStore } from '../../store';
 import { unlock } from '../../lock-unlock';
 
 export function BlockRemovalWarningModal( { rules } ) {
```
Do we still want to keep this component in "block-editor"?

talldan replied:
Ok with moving it, but then we still have the private actions, selectors and reducers for block removal warnings in the block editor package that are all the same family of APIs. I feel like it makes sense to keep those things together.

We could consider gradually moving it all across, but it'd mean changing how this feature integrates with the `removeBlocks` action pretty significantly. Will have a think about whether that's possible.

youknowriad replied:
Ok, let's keep it then :)

youknowriad reviewed (commented):
This is looking good for me.

talldan commented:
Had a chat with @glendaviesnz and have agreed to take this PR over from him as he doesn't have the bandwidth to work on this right now.

talldan committed:
Refactor block removal callback rules

- Callbacks now receive all removed blocks and return the removal prompt message
- Filter for post type in the component instead of the block editor action

talldan committed:
Add in an early return when there are no block removal rules

talldan commented:
I've rebased and refactored some of it that I felt could be improved. I still need to do some testing to make sure I haven't broken anything 😅 . Not sure if there are e2e tests for this feature.

The rules are now an array (to keep the displayed messages in a reliable order) of objects in this shape:
```js
{
    // list of post types the message is displayed for.
    postTypes: [ 'some', 'post', 'types' ],
    // callback 
    callback( removedBlocks ) {
        if ( removedBlocks.some( ( { name } ) => name === 'core/some-block' ) ) {
            return __( 'Message that displays in prompt' );
	}
    }
},
```

So the callback, instead of just checking eligibility for the prompt, now also returns the message itself. I think this gives a bit more flexibility if we want to do some more dynamic messages in the future, plus simplifies the `privateRemoveBlocks` action.

The postType filtering now happens in the component instead of the `privateRemoveBlocks` action.

I'm unsure about the way the messages themselves have changed in this PR, as it seems like there hasn't been much design/copy feedback around it.

talldan committed:
Reflect existing `trunk` behavior of only showing one warning message

talldan committed:
Restore original e2e test

talldan committed:
Fix a few bugs

talldan committed:
Finish sentence

talldan commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,85 @@
+	{
+		// Template blocks.
+		// The warning is only shown when a user manipulates templates or template parts.
+		postTypes: [ 'wp_template', 'wp_template_part' ],
```
Not sure if this message needs to display when editing any other post types than the ones specified here? Or maybe it should be universal (happens for any post type).

talldan commented on packages/editor/src/components/block-removal-warnings/index.js:
```diff
@@ -0,0 +1,85 @@
+			if ( removedTemplateBlocks.length ) {
+				return _n(
+					'Deleting this block will stop your post or page content from displaying on this template. It is not recommended.',
+					'Some of the deleted blocks will stop your post or page content from displaying on this template. It is not recommended.',
```
I revised the copy that was in `trunk` as we don't know that all of the deleted blocks are dangerous to be removed, it could be four paragraphs and one query loop.

We could also go through all the permutations to make sure the messaging is completely correct, but I thought this would be sufficient.

andrewserong reviewed (approved):
LGTM, too! I like the simplified signature for `callback( removedBlocks )`, it feels nice and simple now 👍

Looking good:

<img width="548" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/f4896995-12e8-4301-a6a0-f231a2e3f3e2">

<img width="538" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/6aa03272-f818-4a01-8ea6-6c5e24b95a7a">

talldan committed:
Avoid trying to render `BlockRemovalWarnings` on mobile

talldan committed:
Update test

talldan commented:
Thanks for the reviews ❤️
